src/Capability/Registry/DynamicToolReference.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the official PHP MCP SDK.
+ *
+ * A collaboration between Symfony and the PHP Foundation.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Mcp\Capability\Registry;
+
+use Mcp\Capability\Formatter\ToolResultFormatter;
+use Mcp\Capability\Provider\DynamicToolProviderInterface;
+use Mcp\Schema\Content\Content;
+use Mcp\Schema\Tool;
+use Mcp\Server\ClientAwareInterface;
+use Mcp\Server\ClientGateway;
+
+/**
+ * @author Mateu Aguiló Bosch <mateu@mateuaguilo.com>
+ */
+class DynamicToolReference extends ElementReference implements ClientAwareInterface
+{
+    use DynamicArgumentPreparationTrait;
+
+    public function __construct(
+        public readonly Tool $tool,
+        public readonly DynamicToolProviderInterface $provider,
+        public readonly string $toolName,
+    ) {
+        parent::__construct($this, false);
+    }
+
+    public function setClient(ClientGateway $clientGateway): void
+    {
+        if ($this->provider instanceof ClientAwareInterface) {
+            $this->provider->setClient($clientGateway);
+        }
```

Comment:
Why does a DynamicToolProvider need a `ClientGateway`? Can you give an example? 

It seams like we calling the `DynamicToolProvider` -> Get tool -> Then give ClientInterface to the `DynamicToolProvider`-> then execute the tool. 

This feels wrong. Let's avoid keeping state and move references like this. 

(Shamelessly self-promoting #176)